File: internal/audit/audit.go
// Package audit writes the audit log. Entries never contain secrets.
package audit

import (
	"context"
	"encoding/json"
	"log/slog"

	"github.com/envoryx/envoryx/internal/auth"
	"github.com/envoryx/envoryx/internal/store"
)

// Actions recorded by Envoryx.
const (
	ActionLogin          = "auth.login"
	ActionLoginFailed    = "auth.login_failed"
	ActionLogout         = "auth.logout"
	ActionSetup          = "auth.setup"
	ActionPasswordChange = "auth.password_changed"
	// Rescue CLI (envoryx admin …), run from a shell inside the container.
	ActionPasswordReset   = "auth.password_reset"
	ActionSessionsRevoked = "auth.sessions_revoked"
	ActionTokensRevoked   = "auth.tokens_revoked"
	ActionAccountsReset   = "auth.accounts_reset"

	ActionProjectCreated    = "project.created"
	ActionProjectUpdated    = "project.updated"
	ActionProjectStarted    = "project.started"
	ActionProjectStopped    = "project.stopped"
	ActionProjectRestarted  = "project.restarted"
	ActionProjectDuplicated = "project.duplicated"
	ActionProjectRenamed    = "project.renamed"
	ActionProjectDeleted    = "project.deleted"
	ActionProjectFailed     = "project.failed"

	ActionSettingsChanged = "settings.changed"
	ActionDBToolOpened    = "project.dbtool_opened"
	ActionImageRolledBack = "project.image_rolled_back"
	ActionImageLatest     = "project.image_latest"
	ActionImagesPruned    = "docker.images_pruned"
	ActionOrphansRemoved  = "docker.orphans_removed"

	ActionBackupCreated  = "backup.created"
	ActionBackupRestored = "backup.restored"
	ActionBackupDeleted  = "backup.deleted"

	ActionInstanceBackupCreated  = "instance.backup_created"
	ActionInstanceBackupUploaded = "instance.backup_uploaded"
	ActionInstanceBackupDeleted  = "instance.backup_deleted"
	ActionInstanceRestore        = "instance.restore_scheduled"
	// ActionInstanceRestored is written after a restart applied a scheduled restore. The
	// restore_scheduled entry lives in the database that the restore replaced, so this is
	// the only trace that survives in the restored database.
	ActionInstanceRestored = "instance.restored"

	ActionDBCredentialsViewed = "database.credentials_viewed"
	ActionDBPasswordRotated   = "database.password_rotated"
	ActionDBCreated           = "database.created"
	ActionDBDropped           = "database.dropped"

	ActionTerminalOpened = "terminal.opened"
	ActionExec           = "project.exec"
	ActionRun            = "action.run"

	ActionDeployKeyGenerated = "git.deploy_key_generated"
	ActionGitClone           = "git.clone"
	ActionGitPull            = "git.pull"
	ActionGitCheckout        = "git.checkout"
)

type ipKey struct{}

// WithClientIP stores the request IP in the context so audit entries can record it.
func WithClientIP(ctx context.Context, ip string) context.Context {
	return context.WithValue(ctx, ipKey{}, ip)
}

// Logger appends audit entries.
type Logger struct {
	store *store.Audit
	log   *slog.Logger
}

// New creates an audit logger.
func New(st *store.Audit, log *slog.Logger) *Logger {
	return &Logger{store: st, log: log}
}

// Actor is the user name an audit row shows for a principal: the account, suffixed with
// the API token's name when the request was not a browser session.
func Actor(p auth.Principal) string {
	if p.TokenName != "" {
		return p.Username + " (token: " + p.TokenName + ")"
	}
	return p.Username
}

// Log records an action performed by the principal in ctx (if any). details is
// marshalled to JSON; pass nil for none. Failures to write the audit log are logged but
// do not fail the business operation.
func (l *Logger) Log(ctx context.Context, action, targetType, targetID string, details any) {
	e := store.AuditEntry{Action: action, TargetType: targetType, TargetID: targetID}
	if p, ok := auth.PrincipalFrom(ctx); ok {
		e.UserID, e.Username = p.UserID, Actor(p)
	}
	if ip, ok := ctx.Value(ipKey{}).(string); ok {
		e.IP = ip
	}
	if details != nil {
		b, err := json.Marshal(details)
		if err != nil {
			l.log.Warn("audit details not serialisable", "action", action, "err", err)
		} else {
			e.Details = b
		}
	}
	// Audit writes must succeed even when the request context is already cancelled.
	if err := l.store.Append(context.WithoutCancel(ctx), e); err != nil {
		l.log.Error("audit log write failed", "action", action, "err", err)
	}
	l.log.Info("audit", "action", action, "target", targetType, "id", targetID, "user", e.Username)
}

// LogAs records an action on behalf of an explicit username (e.g. failed logins).
func (l *Logger) LogAs(ctx context.Context, username, action, targetType, targetID string, details any) {
	e := store.AuditEntry{Username: username, Action: action, TargetType: targetType, TargetID: targetID}
	if ip, ok := ctx.Value(ipKey{}).(string); ok {
		e.IP = ip
	}
	if details != nil {
		if b, err := json.Marshal(details); err == nil {
			e.Details = b
		}
	}
	if err := l.store.Append(context.WithoutCancel(ctx), e); err != nil {
		l.log.Error("audit log write failed", "action", action, "err", err)
	}
}

File: web/src/lib/audit.ts
import { formatBytes } from "@/lib/format";
import type { TFunction } from "i18next";
import type { AuditEntry } from "@/api/types";

/** English labels for the backend's audit actions; translated on the client like other keys. */
export const auditActionLabels: Record<string, string> = {
  "auth.login": "Signed in",
  "auth.login_failed": "Sign-in failed",
  "auth.logout": "Signed out",
  "auth.setup": "Initial setup",
  "auth.password_changed": "Password changed",
  "auth.password_reset": "Password reset (rescue command)",
  "auth.sessions_revoked": "All sessions revoked",
  "auth.tokens_revoked": "API tokens revoked",
  "auth.accounts_reset": "Accounts reset (rescue command)",
  "project.created": "Project created",
  "project.updated": "Project updated",
  "project.started": "Project started",
  "project.stopped": "Project stopped",
  "project.restarted": "Project restarted",
  "project.duplicated": "Project duplicated",
  "project.renamed": "Project renamed",
  "project.deleted": "Project deleted",
  "project.failed": "Project creation failed",
  "project.dbtool_opened": "Database browser opened",
  "project.image_rolled_back": "Image rolled back",
  "project.image_latest": "Image set to latest",
  "settings.changed": "Settings changed",
  "docker.images_pruned": "Unused images removed",
  "docker.orphans_removed": "Orphaned resources removed",
  "logs.history_cleared": "Log history deleted",
  "cache.cleared": "Package cache cleared",
  "backup.created": "Backup created",
  "backup.restored": "Backup restored",
  "backup.deleted": "Backup deleted",
  "instance.backup_created": "Instance backup created",
  "instance.backup_uploaded": "Instance backup uploaded",
  "instance.backup_deleted": "Instance backup deleted",
  "instance.restore_scheduled": "Instance restore scheduled",
  "instance.restored": "Instance restored",
  "database.credentials_viewed": "Database credentials viewed",
  "database.password_rotated": "Database password rotated",
  "database.created": "Database created",
  "database.dropped": "Database dropped",
  "database.cloned": "Database cloned",
  "terminal.opened": "Terminal opened",
  "project.exec": "Command run",
  "action.run": "Action run",
  "test.run": "Tests run",
  "cron.run": "Cron job run",
  "action.finished": "Action finished",
  "git.deploy_key_generated": "Deploy key generated",
  "git.clone": "Repository cloned",
  "git.pull": "Repository pulled",
  "git.checkout": "Branch checked out",
  "token.created": "API token created",
  "token.revoked": "API token revoked",
};

export function auditActionLabel(action: string, t: TFunction): string {
  const label = auditActionLabels[action];
  return label ? t(label) : action;
}

/** Who did it: the user, "cli" for rescue commands, Envoryx itself for automatic actions. */
export function auditActor(e: AuditEntry, t: TFunction): string {
  if (e.username) return e.username;
  return t("Envoryx (automatic)");
}

/** A short, readable summary of an entry's details; falls back to the target. */
export function auditDetails(e: AuditEntry, t: TFunction): string {
  const d = e.details ?? {};
  const list = (v: unknown) => (Array.isArray(v) ? v.map(String).join(", ") : "");
  switch (e.action) {
    case "docker.orphans_removed":
      return list(d["removed"]);
    case "docker.images_pruned":
      return t("Images: {{count}}", { count: Number(d["removed"] ?? 0) });
    case "settings.changed":
      return Object.keys(d).join(", ");
    case "cache.cleared":
      return [typeof d["tool"] === "string" ? d["tool"] : "", t("{{size}} freed", { size: formatBytes(Number(d["freed"] ?? 0)) })].filter(Boolean).join(" · ");
    case "project.updated": {
      const changes = d["changes"];
      const keys = changes && typeof changes === "object" ? Object.keys(changes as object).join(", ") : "";
      return typeof d["name"] === "string" ? (keys ? `${d["name"]} · ${keys}` : String(d["name"])) : keys;
    }
    default:
      if (typeof d["name"] === "string") return String(d["name"]);
      return e.targetType ? `${e.targetType} ${e.targetId.slice(0, 8)}` : "—";
  }
}
